done

DeMazGH opened the pull request.

DeMazGH committed:
done

DeMazGH committed:
done

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/FacultyService.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    public Faculty updateFaculty(Faculty faculty) {
+        faculties.put(faculty.getId(), faculty);
```
этот метод отработает как метод добавления. если передать несуществующий id. Здесь надо делать проверку, а есть ли такой id и если нет, то возвращать из метода null, а из контроллера 404

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/StudentService.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    public Student updateStudent(Student student) {
+        students.put(student.getId(), student);
```
этот метод отработает как метод добавления. если передать несуществующий id. Здесь надо делать проверку, а есть ли такой id и если нет, то возвращать из метода null, а из контроллера 404

DeMazGH committed:
fixed update methods

DeMazGH committed:
fixes in tests

DeMazGH committed:
add DB

DeMazGH committed:
done (except for tests)

DeMazGH committed:
add method - findByAgeBetween

DeMazGH committed:
separated sequence

DeMazGH committed:
add find by color and name in Faculty

DeMazGH committed:
SQL query

DeMazGH committed:
add method get students faculty

DeMazGH committed:
add method - getStudentsByFaculty

DeMazGH committed:
small fixes

DeMazGH committed:
add tests for student service

DeMazGH committed:
add tests for FacultyService

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    @GetMapping()
+    public ResponseEntity<Collection<Faculty>> findFacultiesByColor(@RequestParam(required = false) String color,
```
по заданию здесь 1 параметр, а не 2. И лучше его сделать обязательным - зачем делать лишние проверки, если спринг умеет это делать? Т.е. параметр String nameOrColor

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -0,0 +1,72 @@
+        return ResponseEntity.ok(Collections.emptyList());
+    }
+
+    @GetMapping("/getStudentsByFaculty/{facultyId}")
```
правильнее будет /{facultyId}/students

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/StudentController.java:
```diff
@@ -0,0 +1,73 @@
+    public ResponseEntity<Student> findStudent(@PathVariable long studentId) {
+        Student student = studentService.findStudent(studentId);
+        if (student == null) {
+            ResponseEntity.notFound().build();
```
return

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -0,0 +1,72 @@
+    public ResponseEntity<Faculty> findFaculty(@PathVariable long facultyId) {
+        Faculty faculty = facultyService.findFaculty(facultyId);
+        if (faculty == null) {
+            ResponseEntity.notFound().build();
```
return

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -0,0 +1,72 @@
+    @GetMapping("/getStudentsByFaculty/{facultyId}")
+    public ResponseEntity<Collection<Student>> getStudentsByFaculty(@PathVariable long facultyId) {
+        if (facultyService.findFaculty(facultyId) == null) {
+            ResponseEntity.notFound().build();
```
return

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/StudentController.java:
```diff
@@ -0,0 +1,73 @@
+    public ResponseEntity<Faculty> getFacultyByStudentId(@PathVariable long studentId) {
+        Faculty faculty = studentService.getFacultyByStudentId(studentId);
+        if (faculty == null) {
+            ResponseEntity.notFound().build();
```
return

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/StudentController.java:
```diff
@@ -0,0 +1,73 @@
+    public ResponseEntity<Student> updateStudent(@RequestBody Student student) {
+        Student updatedStudent = studentService.updateStudent(student);
+        if (updatedStudent == null) {
+            return ResponseEntity.status(HttpStatus.BAD_REQUEST).build();
```
not found

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/StudentController.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @GetMapping()
+    public ResponseEntity<Collection<Student>> findStudentsFromAge(@RequestParam(required = false) Integer studentAge,
```
не, не нужно было объединять методы. Надо вернуть старый и написать новый, также параметры все должны быть обязательными и не нужны проверки эти

dmitry-bizin commented on src/main/java/ru/hogwarts/school/model/Faculty.java:
```diff
@@ -0,0 +1,77 @@
+            sequenceName = "faculty_sequence",
+            allocationSize = 1)
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "faculty_seq")
+    private long id;
```
Long

dmitry-bizin commented on src/main/java/ru/hogwarts/school/model/Student.java:
```diff
@@ -0,0 +1,85 @@
+            sequenceName = "student_sequence",
+            allocationSize = 1)
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "student_seq")
+    private long id;
```
Long

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/FacultyService.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    public Faculty findFaculty(long id) {
+        return facultyRepository.findById(id).get();
```
здесь исключение будет, это неправильное использование Optional - у тебя просто выкинется NoSuchElementException. Надо делать проверку или вместо get вызывать orElse(null) например

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/FacultyService.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    public Faculty updateFaculty(Faculty faculty) {
+        return facultyRepository.save(faculty);
```
чем этот метод отличается от сохранения? ничем. Здесь надо проверить, что есть сущность с таким id, если нет, то вернуть налл и потом в контроллере 404

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/StudentService.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public Student findStudent(Long id) {
+        return studentRepository.findById(id).get();
```
здесь исключение будет, это неправильное использование Optional - у тебя просто выкинется NoSuchElementException. Надо делать проверку или вместо get вызывать orElse(null) например

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/StudentService.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public Student updateStudent(Student student) {
+        return studentRepository.save(student);
```
чем этот метод отличается от сохранения? ничем. Здесь надо проверить, что есть сущность с таким id, если нет, то вернуть налл и потом в контроллере 404

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/StudentService.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public Collection<Student> findByAge(Integer age) {
+        return findAllStudents().stream().filter(student -> Objects.equals(student.getAge(), age)).toList();
```
нет, это неверно. Никогда не надо фильтровать в коде. Вот будет у тебя в БД сотни тысяч строк и здесь будет OutOfMemoryError, потому что в хип загрузил сотни тысяч объектов. Надо использовать метод репозитория

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/StudentService.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public Faculty getFacultyByStudentId(long studentId) {
+        return studentRepository.findStudentById(studentId).getFaculty();
```
здесь если не будет студента по id, то вернётся налл и будет исключение - нужна проверка

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/FacultyService.java:
```diff
@@ -0,0 +1,53 @@
+        return facultyRepository.findAll();
+    }
+
+    public Collection<Faculty> findByColor(String color) {
```
метод должен принимать 1 параметр - String nameOrColor и вызывать метод репозитория с двумя параметрами и передавать в него этот параметр 2 раза

dmitry-bizin commented on src/script.sql:
```diff
@@ -0,0 +1,6 @@
+select * from student;
+select * from student where age > 10 and age < 17;
```
лучше использовать between

dmitry-bizin commented on src/test/java/ru/hogwarts/school/service/StudentServiceTest.java:
```diff
@@ -0,0 +1,106 @@
+
+//    @Test
+//    void shouldReturnStudentInMethodFind() {
+//        when(studentRepositoryMock.findById(LONG_NUM_2).get()).thenReturn(STUDENT_BBB);
```
здесь надо сделать так, чтобы метод findById возвращал Optional и не надо get вызывать

dmitry-bizin commented on src/test/java/ru/hogwarts/school/service/FacultyServiceTest.java:
```diff
@@ -0,0 +1,103 @@
+
+//    @Test
+//    void shouldReturnFacultyInMethodFindFaculty() {
+//        when(facultyRepositoryMock.findById(LONG_NUM_1).get()).thenReturn(FACULTY_AAA);
```
здесь надо сделать так, чтобы метод findById возвращал Optional и не надо get вызывать

DeMazGH committed:
fixed bags

DeMazGH committed:
add model avatar

DeMazGH committed:
add avatar Controller Service Repository

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -0,0 +1,75 @@
+
+    @GetMapping("/findByColor")
+    public ResponseEntity<Collection<Faculty>> findFacultiesByColor(@RequestParam String color) {
+        if (color != null && !color.isBlank()) {
```
проверки можно убрать, т.к. параметр обязательный и проверки лишние здесь

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -0,0 +1,75 @@
+
+    @GetMapping("/findByColorOrName")
+    public ResponseEntity<Collection<Faculty>> findFacultiesByColorOrName(@RequestParam String nameOrColor) {
+        if (nameOrColor != null && !nameOrColor.isBlank()) {
```
проверки можно убрать, т.к. параметр обязательный и проверки лишние здесь

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/StudentController.java:
```diff
@@ -0,0 +1,77 @@
+
+    @GetMapping()
+    public ResponseEntity<Collection<Student>> findStudentsFromAge(@RequestParam Integer studentAge) {
+        if (studentAge != null) {
```
проверку можно убрать, т.к. параметр обязательный и проверка лишняя здесь

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/StudentController.java:
```diff
@@ -0,0 +1,77 @@
+    @GetMapping("/findStudentByAgeInInterval")
+    public ResponseEntity<Collection<Student>> findStudentsFromAgeInTheInterval(@RequestParam Integer minAge,
+                                                                                @RequestParam Integer maxAge) {
+        if (minAge != null && maxAge != null && minAge >= 0 && minAge < maxAge) {
```
проверки можно убрать, т.к. параметр обязательный и проверки лишние здесь

DeMazGH committed:
3.5 done

DeMazGH committed:
small fixes

DeMazGH committed:
small fixes

DeMazGH committed:
add StudentControllerTest

DeMazGH committed:
add 2 constructor is in method Student

DeMazGH committed:
add tests from SC (not all works)

DeMazGH committed:
add tests from FC

DeMazGH committed:
small fixes

DeMazGH committed:
small fixes

DeMazGH committed:
done

dmitry-bizin commented on src/test/java/ru/hogwarts/school/controller/StudentControllerTest.java:
```diff
@@ -0,0 +1,244 @@
+        thenStudentHasBeenCreated(createResponse);
+        student = createResponse.getBody();
+
+//        if (createResponse.getBody() != null) {
```
это можно проверить как раз через ассёрты. Т.е. `Assertions.assertThat(createResponse.getBody()).isNotNull()`

dmitry-bizin commented on src/test/java/ru/hogwarts/school/controller/StudentControllerTest.java:
```diff
@@ -0,0 +1,244 @@
+
+    private void thenFacultyHasBeenFind(ResponseEntity<Faculty> response) {
+        Assertions.assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
+        Assertions.assertThat(response.getBody()).isNotNull();
```
здесь лучше бы не просто что non null, а ещё проверять сами поля, так проверки будут качественнее

dmitry-bizin commented on src/test/java/ru/hogwarts/school/controller/StudentControllerTest.java:
```diff
@@ -0,0 +1,244 @@
+
+    private void thenStudentHasBeenCreated(ResponseEntity<Student> response) {
+        Assertions.assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
+        Assertions.assertThat(response.getBody()).isNotNull();
```
здесь лучше бы не просто что non null, а ещё проверять сами поля, так проверки будут качественнее

dmitry-bizin commented on src/test/java/ru/hogwarts/school/service/StudentServiceTest.java:
```diff
@@ -0,0 +1,121 @@
+
+        when(studentRepositoryMock.findByAge(NUM_1)).thenReturn(actual);
+
+        actual = (List<Student>) out.findByAge(NUM_1);
```
а зачем здесь каст?

dmitry-bizin replied:
если findByAge возвращает коллекцию, то значит надо actual сделать тоже типа Collection

DeMazGH committed:
add 3 endpoints from StudentController

DeMazGH committed:
small fixes in tests

DeMazGH committed:
add tests from 3 endpoints:
- getFiveLatestStudents
- getAverageAgeOfAllStudents
- getAmountOfAllStudents

DeMazGH committed:
add endpoint in AvatarController^
- findByPagination

DeMazGH committed:
small fixes

DeMazGH committed:
homework 4.2 one

DeMazGH committed:
homework 4.3

dmitry-bizin commented on src/main/resources/liquibase/changelog-master.yml:
```diff
@@ -0,0 +1,3 @@
+databaseChangeLog:
+  include:
```
здесь перед include нужно тире и пробел после него

DeMazGH committed:
homework 4.3

DeMazGH committed:
homework 4.3 done

DeMazGH committed:
4.4
add logs in services

DeMazGH committed:
4.4
errors from build

DeMazGH committed:
4.5
add method - getStudentsWithNameIsStartsFromA
add tests from this method

DeMazGH committed:
4.5
fix method to getStudentsNameIsStartsFromA
fix tests from this method

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/AvatarService.java:
```diff
@@ -0,0 +1,78 @@
+//@Transactional
+public class AvatarService {
+
+    Logger logger = LoggerFactory.getLogger(AvatarService.class);
```
private final

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/FacultyService.java:
```diff
@@ -0,0 +1,69 @@
+    private final FacultyRepository facultyRepository;
+    private final StudentRepository studentRepository;
+
+    Logger logger = LoggerFactory.getLogger(FacultyService.class);
```
private final

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/StudentService.java:
```diff
@@ -0,0 +1,95 @@
+
+    private final StudentRepository studentRepository;
+
+    Logger logger = LoggerFactory.getLogger(StudentService.class);
```
private final

DeMazGH committed:
4.4
small fixes

DeMazGH committed:
4.5
- add endpoint - getAverageAgeOfAllStudentsWithStreams
- add test from its

DeMazGH committed:
4.5
- add logger in StudentService
- add endpoint - getLongestFacultyName in FacultyController
- add tests from method - getLongestFacultyName

DeMazGH committed:
4.5
- add endpoint - getSum in InfoController
- add tests from its

DeMazGH committed:
4.5 done

DeMazGH committed:
4.6 done

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/StudentService.java:
```diff
@@ -100,4 +100,42 @@ public Double getAverageAgeOfAllStudentsWithStreams() {
+        }).start();
+    }
+
+    private synchronized void printStudentName(Student student) {
```
порядка на самом деле нет - позапускай метод несколько раз и ты увидишь, что вывод будет в разном порядке. Чтобы синхронизация обеспечивала порядок, можно её сделать вот как. Завести 2 поля - счётчик и Object, синхронизация будет на Object поле, а метод принимает список, из списка достаёт студента по счётчику, выводит и увеличивает счётчик на 1

DeMazGH committed:
4.6
fix synchronized method

DeMazGH committed:
Create README.md
